packages/babel-helpers/src/helpers.js

```diff
@@ -270,6 +270,170 @@ helpers.createClass = template(`
   })()
 `);
 
+//NOTE: convention is 'descriptor' is used for element descriptors, while 'propertyDescriptor' is used for
+//property descriptor in all helpers related to decorators-2
+helpers.decorate = template(`
+  (function (constructor, undecorated, memberDecorators, heritage) {
+    const prototype = constructor.prototype;
+    let finishers = [];
+    const elementDescriptors = new Map(); // elementDescriptors is meant to be an array, so this will be converted later
+
+    for (const [key, isStatic] of undecorated) {
+      const target = isStatic ? constructor : prototype;
+      const propertyDescriptor = Object.getOwnPropertyDescriptor(target, key);
+      elementDescriptors.set(
+        [key, isStatic], 
+        babelHelpers.makeElementDescriptor(
+          "property",
+          key,
+          isStatic,
+          propertyDescriptor,
+        )
+      );
+    }
+
+    // decorate and store in elementDescriptors
+    for (const [key, decorators, isStatic] of memberDecorators) {
+      const target = isStatic ? constructor : prototype;
+      const propertyDescriptor =
+        elementDescriptors.has([key, isStatic]) && elementDescriptors.get([key, isStatic]).descriptor 
+        || Object.getOwnPropertyDescriptor(target, key);
+
+      const elementDescriptor = babelHelpers.makeElementDescriptor(
+        "property",
+        key,
+        isStatic,
+        propertyDescriptor,
+      )
+      const decorated = babelHelpers.decorateElement(elementDescriptor, decorators);
+
+      elementDescriptors.set([key, isStatic], decorated.descriptor);
+
+      for (const extra of decorated.extras) {
+        // extras is an array of element descriptors
+        elementDescriptors.set([extra.key, extra.isStatic], extra);
+      }
+
+      finishers = finishers.concat(decorated.finishers);
+    }
+
+    return function(classDecorators) {
+      const result = babelHelpers.decorateClass(
+        constructor,
+        classDecorators,
+        heritage,
+        Array.from(elementDescriptors.values())
+      );
+
+      finishers = finishers.concat(result.finishers);
+      //TODO: heritage hacks so result.constructor has the correct prototype and instanceof results
+      //TODO: step 38 and 39, what do they mean "initialize"?
```

Comment:
"Initialize" was about having methods and the constructor start in some state where they would be unusable, and make them usable later. In [my follow-on decorators proposal](https://github.com/littledan/proposal-unified-class-features), I'm suggesting to get rid of this initialized/uninitialized state. I think it's OK to go without it to start, even if it's possible that TC39 will ask to bring it back.